src/routes/session.rs

```diff
@@ -101,3 +105,139 @@ pub fn delete(mut cookies: Cookies) -> Redirect {
     }
     Redirect::to("/")
 }
+
+#[derive(Clone)]
+pub struct ResetRequest {
+    pub mail: String,
+    pub id: String,
+    pub creation_date: Instant,
+}
+
+impl PartialEq for ResetRequest {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+#[get("/password-reset")]
+pub fn password_reset_request_form(conn: DbConn, intl: I18n) -> Ructe {
+    render!(session::password_reset_request(
+        &(&*conn, &intl.catalog, None),
+        &ResetForm::default(),
+        ValidationErrors::default()
+    ))
+}
+
+#[derive(FromForm, Validate, Default)]
+pub struct ResetForm {
+    #[validate(email)]
+    pub email: String,
+}
+
+#[post("/password-reset", data = "<form>")]
+pub fn password_reset_request(
+    conn: DbConn,
+    intl: I18n,
+    mail: State<Arc<Mutex<Mailer>>>,
+    form: Form<ResetForm>,
+    requests: State<Arc<Mutex<Vec<ResetRequest>>>>
+) -> Ructe {
+    if User::find_by_email(&*conn, &form.email).is_ok() && requests.lock().unwrap().iter().find(|x| x.mail == form.email.clone()).is_none() {
```

Comment:
You should maintain the lock instead of locking and unlocking many times

Reply:
I can't because I need to borrow it both mutably and immutably and Rust won't let me do that. But maybe there is a solution to do it I didn't find?

Reply:
I've removed a lot of code, but here is a poc of how I would do it (and some other things, trashing old things sooner and don't use the feature) https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=2c32af184c99ff62ee99f94aa2e4fa13
Sadly I can't compile this branch so I can't give you a clean commit, but it should put you on the right track